python/tvm/contrib/debugger/debug_runtime.py

```diff
@@ -0,0 +1,327 @@
+"""Graph debug runtime executes TVM debug packed functions."""
+
+import os
+import tempfile
+import shutil
+from tvm import ndarray as nd
+from tvm._ffi.base import string_types
+from tvm.contrib import graph_runtime
+from tvm._ffi.function import get_global_func
+from . import common
+from . import debug_result
+
+_DUMP_ROOT_PREFIX = "tvmdbg_"
+_DUMP_PATH_PREFIX = "_tvmdbg_"
+
+def create(graph_json_str, libmod, ctx, dbg_ux=None, dump_root=None):
+    """Create a runtime executor module given a graph and module.
+
+    Parameters
+    ----------
+    graph_json_str : str or graph class
+        The graph to be deployed in json format output by nnvm graph.
+        The graph can only contain one operator(tvm_op) that
+        points to the name of PackedFunc in the libmod.
+
+    libmod : tvm.Module
+        The module of the corresponding function.
+
+    ctx : TVMContext
+        The context to deploy the module, can be local or remote.
+
+    dbg_ux : str
+        To select which ux user needs, Example, curses/tensorboard/None.
+        None will just do the dumping
+
+    dump_root : str
+        To select which folder the outputs should be kept.
+        None will make a temp folder in /tmp/tvmdbg<rand_string> and does the dumping
+    Returns
+    -------
+    graph_module : GraphModuleDebug
+        Debug Runtime graph module that can be used to execute the graph.
+    """
+    if not isinstance(graph_json_str, string_types):
+        try:
+            graph_json_str = graph_json_str._tvm_graph_json()
+        except AttributeError:
+            raise ValueError("Type %s is not supported" % type(graph_json_str))
+    device_type = ctx.device_type
+    device_id = ctx.device_id
+    try:
+        fcreate = get_global_func("tvm.graph_runtime_debug.create")
+    except ValueError:
+        raise ValueError("Please set '(USE_GRAPH_RUNTIME_DEBUG ON)' in " \
+                         "config.cmake and rebuild TVM to enable debug mode")
+    func_obj = fcreate(graph_json_str, libmod, device_type, device_id)
+    return GraphModuleDebug(func_obj, ctx, graph_json_str, dbg_ux, dump_root)
+
+
+class GraphModuleDebug(graph_runtime.GraphModule):
+    """Graph debug runtime module.
+
+    This is a debug wrapper over the TVM runtime.
+    Runtime interfaces are wrapped with debug functionalities.
+    Manage the debug framework to format the debug data and
+    trigger the user interfaces.
+
+    Parameters
+    ----------
+    module : Module
+        The interal tvm module that holds the actual graph functions.
+
+    ctx : TVMContext
+        The context this module is under.
+
+    graph_json_str : str or graph class
+        The graph to be deployed in json format output by nnvm graph.
+        The graph can only contain one operator(tvm_op) that
+        points to the name of PackedFunc in the libmod.
```

Comment:
graph json definition is detailed above already (tvmdbg.rst), just keep it simple.

"Content of graph json file in string format" or similar.